### PR TITLE
Proximity API

### LINKED ISSUE #14: Implement Proximity API
Implement the API for the [Proximity](https://github.com/tock/tock/blob/master/capsules/extra/src/proximity.rs)

### PR DESCRIPTION
**Pull request by alexandruCarp**
Started working to resolve #14. Opened the draft PR to assign myself this task.

### CODE REVIEW

**Comment by alexandruCarp** [comment]
I have implemented the API and made an example app to test it, but I found out that the microbit does not have a proximity sensor. What else should I use for testing?

File: apis/proximity/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    /// Initiate a synchronous on_interrupt proximity measurement.
+    /// Returns Ok(proximity_value) if the operation was successful
+    /// proximity_value is in [0, 255] range,
+    /// where '255' indicates the closest measurable distance and '0' that no object is detected
+    /// Returns only when proximity_value < lower or proximity_value > upper
+    /// lower - lower interrupt threshold for sensor --> range is [0,255]
+    /// upper - upper interrupt threshold for sensor --> range is [0,255]
+    /// lower <= upper
+    pub fn read_proximity_on_interrupt_sync(lower: u32, upper: u32) -> Result<u8, ErrorCode> {
```
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:75]:
```suggestion
    pub fn wait_for_value_between(lower: u32, upper: u32) -> Result<u8, ErrorCode> {
```

Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:75]:
```suggestion
    pub fn wait_for_value_between(lower: u32, upper: u32) -> Result<u8, ErrorCode> {
```

File: apis/proximity/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        }
+    }
+
+    /// Initiate an on_interrupt proximity measurement
+    /// Executes the callback only if
+    /// proximity_value < lower or proximity_value > upper
+    /// lower - lower interrupt threshold for sensor --> range is [0,255]
+    /// upper - upper interrupt threshold for sensor --> range is [0,255]
+    /// lower <= upper
+    pub fn read_proximity_on_interrupt(lower: u32, upper: u32) -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:60]: You have to explain what `on_interupt` means. Does it 
- read values continuously and reports the value only if it is in between `lower` and `upper`
- reads value on a certain external interrupt and reports it only if it is in between `lower` and `upper`

Put a link to the capsule's documentation in the comments if you cannot figure it out from the documentation.

Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:60]:
```suggestion
    pub fn read_proximity_on_interrupt(lower: u8, upper: u8) -> Result<(), ErrorCode> {
```

This makes the `> 255` check unnecessary.

File: apis/proximity/src/lib.rs (hunk 1 above)
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:75]:
```suggestion
    pub fn read_proximity_on_interrupt_sync(lower: u8, upper: u8) -> Result<u8, ErrorCode> {
```

File: examples/proximity.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    loop {
+        match Proximity::read_proximity_sync() {
+            Ok(prox_val) => writeln!(Console::writer(), "Proximity: {}\n", prox_val).unwrap(),
+            Err(_) => writeln!(Console::writer(), "error while reading proximity",).unwrap(),
+        }
+
+        Alarm::sleep_for(Milliseconds(2000)).unwrap();
+    }
+}
```
Comment (alexandruradovici) [inline examples/proximity.rs:34]: Move this in the `Ok(_)` branch.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
Please run `cargo fmt` to format the source.

File: apis/proximity/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Initiate a proximity measurement
+    pub fn read_proximity() -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:30]:
```suggestion
    pub fn read() -> Result<(), ErrorCode> {
```

File: apis/proximity/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        S::command(DRIVER_NUM, READ, 0, 0).to_result()
+    }
+
+    /// Initiate a synchronous proximity measurement.
+    /// Returns Ok(proximity_value) if the operation was successful
+    /// proximity_value is in [0, 255] range,
+    /// where '255' indicates the closest measurable distance and '0' that no object is detected
+    pub fn read_proximity_sync() -> Result<u8, ErrorCode> {
```
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:38]:
```suggestion
    pub fn read_sync() -> Result<u8, ErrorCode> {
```

File: apis/proximity/src/lib.rs (hunk 2 above)
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:60]:
```suggestion
    pub fn read_on_interrupt(lower: u32, upper: u32) -> Result<(), ErrorCode> {
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/proximity/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    S::yield_wait();
+                }
+            }
+        });
+
+        match listener.get() {
+            None => Err(ErrorCode::Busy),
+            Some(proximity) => Ok(proximity.0 as u8),
+        }
+    }
```
Comment (alexandruradovici) [inline apis/proximity/src/lib.rs:89]: I think you have too loop until the upcall was called.

File: apis/proximity/src/tests.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    let driver = fake::Proximity::new();
+    kernel.add_driver(&driver);
+
+    assert_eq!(Proximity::read(), Ok(()));
+    assert_eq!(Proximity::read(), Err(ErrorCode::Busy));
+    assert_eq!(Proximity::read_on_interrupt(0, 0), Err(ErrorCode::Busy));
+
+    driver.set_value(100);
+
+    assert_eq!(Proximity::read_on_interrupt(0, 0), Ok(()));
```
Comment (alexandruradovici) [inline apis/proximity/src/tests.rs:34]: Why is this `Ok(())` here?

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/proximity/src/tests.rs (hunk 7 above)
Comment (alexandruCarp) [reply apis/proximity/src/tests.rs:34]: `driver.set_value(100)` executes a callback, consuming the read initiated by `read` at [line 28](https://github.com/alexandruCarp/libtock-rs/blob/7ce4c05c3fd2f4e718b3d4d6f9da42c897b8a447/apis/proximity/src/tests.rs#L28), so it is ok to initiate another measurement, be it on_interrupt or a normal read.

Comment (alexandruradovici) [reply apis/proximity/src/tests.rs:34]: Ok

**Review by alexandruradovici — DISMISSED** [review]
Looks good, send it to upstream and place a link to it.

**Comment by alexandruCarp** [comment]
Opened  PR tock/libtock-rs#462

**Comment by alexandruradovici** [comment]
Merged into master.

**Pull request closed without merging**